hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3336,6 +3338,16 @@ private TableGroup consumeAttributeJoin(
 			registerEntityNameProjectionUsage( sqmJoin, getActualTableGroup( joinedTableGroup, sqmJoin ) );
 		}
 		registerPathAttributeEntityNameUsage( sqmJoin, ownerTableGroup );
+		if ( !sqmJoin.hasTreats() && sqmJoin.getNodeType().getSqmPathType() instanceof EntityDomainType ) {
+			final EntityDomainType<?> entityDomainType = (EntityDomainType<?>) sqmJoin.getNodeType().getSqmPathType();
+			final String entityName = entityDomainType.getHibernateEntityName();
+			final EntityPersister entityDescriptor = domainModel.findEntityDescriptor( entityName );
+			if ( entityDescriptor != null && entityDescriptor.getSuperMappingType() != null ) {
+				// This is a non-treated join with a subtype entity, register a TREAT entity name use
+				// to filter only the correctly typed records
```

Comment:
Can you also add an explanation to the code comment about why this is necessary?

Reply:
> to filter only the correctly typed records

This is why it's necessary. I've expanded the comment a bit more but lmk if it's not clear enough.